pyccel/codegen/printing/ccode.py

```diff
@@ -306,24 +335,43 @@ def copy_NumpyArray_Data(self, expr):
         lhs = expr.lhs
         if rhs.rank == 0:
             raise NotImplementedError(str(expr))
+        order = lhs.order
+        transpose_arg = None
         dummy_array_name = self.scope.get_new_name('array_dummy')
         declare_dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
         dtype = self.find_in_ndarray_type_registry(self._print(rhs.dtype), rhs.precision)
         arg = rhs.arg if isinstance(rhs, NumpyArray) else rhs
+        if isinstance(arg, PythonList) and not isinstance(arg[0], Variable) and order=="F":
+            import numpy as np
+            transpose_arg = list(np.transpose(arg))
```

Comment:
Yes, I was skeptical at first as well, but it worked:
This line for example: ```a = np.array([[1, 2, 4], [5, 6, 7]], order="F")```
Printing its (type, value): ```<class 'pyccel.ast.builtins.PythonList'>, ((1, 2, 4), (5, 6, 7))```
Printing its list(transpose): ```[array([Literal(1), Literal(5)], dtype=object), array([Literal(2), Literal(6)], dtype=object), array([Literal(4), Literal(7)], dtype=object)] this is transpose arg```

Reply:
I suspect this might cause infinite loops if the size of the `PythonList` is unknown, but I'm not sure if I can think up an example. Maybe:
```
n = 3
m = 4
a = np.array([[1]*n]*m, order="F")
```
But apparently pyccel doesn't handle this anyway.

I assume `transpose` is using `PythonList.__getitem__` or `PythonList.__iter__`

Or anything where `is_homogeneous = True`, but that already seems to be the case here

Reply:
You need to make sure you write a few tests for this to make sure it definitely doesn't cause infinite loops

Reply:
Actually, it did cause inifinite loops, I had to add these (kinda ugly) lines:
```python
while isinstance(transpose_arg[0], (np.ndarray, list)):
       arg = [list(i) if isinstance(i, np.ndarray) else i for i in ```self._flatten_list(transpose_arg)]
       transpose_arg = arg
```
I will try to break it with some tests, just to be sure.

Reply:
Huh, that's not where I expected the infinite loop :sweat_smile:

Reply:
Oh, then I must've not understood what you meant, the infinite loop would occure in numpy.transpose?

Reply:
Yes. `__iter__` usually uses `__getitem__` it does `a[0]`, `a[1]`, etc. until an `IndexError` is raised. But if the size of the element is unknown then we can go on forever as we know the result (a `IndexedElement('a', 14352354)`). If we are not using a function which can raise an `IndexError` then we will get an infinite loop

Reply:
Okey, if I am understanding you correctly, am seeing two potential options, creating my own transpose function that will take this into consideration, or add something to the class ```PythonTuple```
How would we raise ```IndexError``` if the size is unknown anyway.

Reply:
Actually, I can just remove the transpose and try to change the ```order``` in C,  like I do with ```Variables```.

Reply:
Possibly you just need to look into the `PythonTuple` class to check if this can happen or not. I don't know if you can have a `PythonTuple` of unknown size. The only ways I can think of to create one involve extra classes.

> How would we raise `IndexError` if the size is unknown anyway.

You can't

> Actually, I can just remove the transpose and try to change the order in C, like I do with Variables.

That should definitely work for homogenous tuples. I think inhomogeneous tuple always have a known size. So maybe it's just a case of treating the 2 differently?

Reply:
Hmm, for some reason I tought inhomogeneous tuples are not supported in Pyccel, so their might be some bugs there, I need to check.
> So maybe it's just a case of treating the 2 differently?
OK, that would work!

Reply:
@EmilyBourne Transpose is no longer in use, I just copy them in `C` using `t_ndarray` strides!
The function `array_copy_data` might need some optimizations but it is working now.